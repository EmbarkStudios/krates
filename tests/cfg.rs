mod util;

use krates::cfg_expr::targets;
use util::{build, cmp};

#[test]
fn ignores_non_linux() {
    let mut kb = krates::Builder::new();
    kb.include_targets(targets::ALL_BUILTINS.iter().filter_map(|ti| {
        if ti.os == Some(targets::Os::linux) {
            Some((ti.triple.clone(), Vec::new()))
        } else {
            None
        }
    }));

    let grafs = build("all-features.json", kb).unwrap();

    let targets: Vec<_> = targets::ALL_BUILTINS
        .iter()
        .filter(|ti| ti.os == Some(targets::Os::linux))
        .collect();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[test]
fn ignores_non_tier1() {
    let mut kb = krates::Builder::new();

    let targets = vec![
        targets::get_builtin_target_by_triple("i686-pc-windows-gnu").unwrap(),
        targets::get_builtin_target_by_triple("i686-pc-windows-msvc").unwrap(),
        targets::get_builtin_target_by_triple("i686-unknown-linux-gnu").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-apple-darwin").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-pc-windows-gnu").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-pc-windows-msvc").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-unknown-linux-gnu").unwrap(),
    ];

    kb.include_targets(targets.iter().map(|ti| (ti.triple.clone(), vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[test]
fn ignores_non_wasm() {
    let mut kb = krates::Builder::new();

    let targets = vec![targets::get_builtin_target_by_triple("wasm32-unknown-unknown").unwrap()];

    kb.include_targets(targets.iter().map(|ti| (ti.triple.clone(), vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[cfg(feature = "targets")]
#[test]
fn handles_non_builtin() {
    let mut kb = krates::Builder::new();

    use krates::target_lexicon::Triple;

    let xbox: Triple = "x86_64-xboxone-windows-msvc".parse().unwrap();

    kb.include_targets(std::iter::once(("x86_64-xboxone-windows-msvc", vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    !expr.eval(|pred| match pred {
                        krates::cfg_expr::Predicate::Target(tp) => tp.matches(&xbox),
                        _ => false,
                    })
                } else {
                    "x86_64-xboxone-windows-msvc" != cfg
                }
            } else {
                false
            }
        },
    );
}
