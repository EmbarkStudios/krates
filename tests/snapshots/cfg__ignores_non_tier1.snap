---
source: tests/cfg.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 (path+file:///home/jake/code/krates/tests/ws/a)" ]
    1 [ label = "crate b 0.1.0 (path+file:///home/jake/code/krates/tests/ws/b)" ]
    2 [ label = "crate bindgen 0.59.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    3 [ label = "crate bitflags 1.3.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    4 [ label = "crate c 0.1.0 (path+file:///home/jake/code/krates/tests/ws/c)" ]
    5 [ label = "crate cc 1.0.73 (git+https://github.com/alexcrichton/cc-rs#53fb72c87e5769a299f1886ead831901b9c775d6)" ]
    6 [ label = "crate cc 1.0.73 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate cexpr 0.6.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate cfg-if 0.1.10 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate cfg-if 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate clang-sys 1.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate coreaudio-rs 0.9.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate coreaudio-sys 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate difference 2.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate glob 0.3.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    15 [ label = "crate lazy_static 1.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "crate lazycell 1.3.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    17 [ label = "crate leftpad 0.2.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    18 [ label = "crate libc 0.2.133 (registry+https://github.com/rust-lang/crates.io-index)" ]
    19 [ label = "crate libloading 0.7.3 (registry+https://github.com/rust-lang/crates.io-index)" ]
    20 [ label = "crate memchr 2.5.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    21 [ label = "crate minimal-lexical 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    22 [ label = "crate nix 0.16.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    23 [ label = "crate nom 7.1.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    24 [ label = "crate once_cell 1.15.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    25 [ label = "crate peeking_take_while 0.1.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    26 [ label = "crate proc-macro2 1.0.44 (registry+https://github.com/rust-lang/crates.io-index)" ]
    27 [ label = "crate quote 1.0.21 (registry+https://github.com/rust-lang/crates.io-index)" ]
    28 [ label = "crate regex 1.6.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    29 [ label = "crate regex-syntax 0.6.27 (registry+https://github.com/rust-lang/crates.io-index)" ]
    30 [ label = "crate ring 0.16.20 (registry+https://github.com/rust-lang/crates.io-index)" ]
    31 [ label = "crate rustc-hash 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    32 [ label = "crate shlex 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    33 [ label = "crate spin 0.5.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    34 [ label = "crate unicode-ident 1.0.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    35 [ label = "crate untrusted 0.7.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    36 [ label = "crate void 1.0.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    37 [ label = "crate winapi 0.2.8 (registry+https://github.com/rust-lang/crates.io-index)" ]
    38 [ label = "crate winapi 0.3.9 (registry+https://github.com/rust-lang/crates.io-index)" ]
    39 [ label = "crate winapi-i686-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    40 [ label = "crate winapi-x86_64-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    41 [ label = "feature default" ]
    42 [ label = "feature default" ]
    43 [ label = "feature default" ]
    44 [ label = "feature clang_6_0" ]
    45 [ label = "feature runtime" ]
    46 [ label = "feature std" ]
    47 [ label = "feature unicode" ]
    48 [ label = "feature default" ]
    49 [ label = "feature default" ]
    50 [ label = "feature default" ]
    51 [ label = "feature default" ]
    52 [ label = "feature std" ]
    53 [ label = "feature audio_toolbox" ]
    54 [ label = "feature audio_unit" ]
    55 [ label = "feature core_audio" ]
    56 [ label = "feature core_midi" ]
    57 [ label = "feature open_al" ]
    58 [ label = "feature runtime" ]
    59 [ label = "feature errhandlingapi" ]
    60 [ label = "feature libloaderapi" ]
    61 [ label = "feature default" ]
    62 [ label = "feature extra_traits" ]
    63 [ label = "feature default" ]
    64 [ label = "feature std" ]
    65 [ label = "feature std" ]
    66 [ label = "feature proc-macro" ]
    67 [ label = "feature unicode" ]
    68 [ label = "feature unicode-age" ]
    69 [ label = "feature unicode-bool" ]
    70 [ label = "feature unicode-case" ]
    71 [ label = "feature unicode-gencat" ]
    72 [ label = "feature unicode-perl" ]
    73 [ label = "feature unicode-script" ]
    74 [ label = "feature unicode-segment" ]
    75 [ label = "feature std" ]
    76 [ label = "feature ntsecapi" ]
    77 [ label = "feature wtypesbase" ]
    78 [ label = "feature leftpad" ]
    79 [ label = "feature leftier-strings" ]
    80 [ label = "feature lazy_static" ]
    81 [ label = "feature libloading" ]
    82 [ label = "feature clang_5_0" ]
    83 [ label = "feature clang_4_0" ]
    84 [ label = "feature clang_3_9" ]
    85 [ label = "feature clang_3_8" ]
    86 [ label = "feature clang_3_7" ]
    87 [ label = "feature clang_3_6" ]
    88 [ label = "feature clang_3_5" ]
    89 [ label = "feature audio_toolbox" ]
    90 [ label = "feature audio_unit" ]
    91 [ label = "feature core_audio" ]
    92 [ label = "feature open_al" ]
    93 [ label = "feature core_midi" ]
    94 [ label = "feature std" ]
    95 [ label = "feature alloc" ]
    96 [ label = "feature alloc" ]
    97 [ label = "feature race" ]
    98 [ label = "feature unicode-age" ]
    99 [ label = "feature unicode-bool" ]
    100 [ label = "feature unicode-case" ]
    101 [ label = "feature unicode-gencat" ]
    102 [ label = "feature unicode-perl" ]
    103 [ label = "feature unicode-script" ]
    104 [ label = "feature unicode-segment" ]
    105 [ label = "feature alloc" ]
    106 [ label = "feature dev_urandom_fallback" ]
    107 [ label = "feature once_cell" ]
    108 [ label = "feature std" ]
    109 [ label = "feature std" ]
    110 [ label = "feature std" ]
    0 -> 1 [ label = "" ]
    0 -> 41 [ label = "(dev)" ]
    0 -> 41 [ label = "(build) 'cfg(target_os = \"linux\")'" ]
    1 -> 41 [ label = "" ]
    1 -> 42 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    2 -> 43 [ label = "" ]
    2 -> 7 [ label = "" ]
    2 -> 44 [ label = "" ]
    2 -> 45 [ label = "" ]
    2 -> 15 [ label = "" ]
    2 -> 16 [ label = "" ]
    2 -> 25 [ label = "" ]
    2 -> 26 [ label = "" ]
    2 -> 27 [ label = "" ]
    2 -> 46 [ label = "" ]
    2 -> 47 [ label = "" ]
    2 -> 48 [ label = "" ]
    2 -> 49 [ label = "" ]
    4 -> 5 [ label = "(build)" ]
    4 -> 50 [ label = " 'x86_64-apple-darwin'" ]
    4 -> 51 [ label = "(dev)" ]
    4 -> 15 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    4 -> 17 [ label = "" ]
    4 -> 18 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    4 -> 22 [ label = " 'x86_64-unknown-linux-gnu'" ]
    4 -> 33 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    4 -> 37 [ label = " 'cfg(target_os = \"windows\")'" ]
    7 -> 52 [ label = "" ]
    10 -> 14 [ label = "" ]
    10 -> 14 [ label = "(build)" ]
    10 -> 18 [ label = "" ]
    10 -> 19 [ label = "" ]
    11 -> 43 [ label = "" ]
    11 -> 12 [ label = "" ]
    11 -> 53 [ label = "" ]
    11 -> 54 [ label = "" ]
    11 -> 55 [ label = "" ]
    11 -> 56 [ label = "" ]
    11 -> 57 [ label = "" ]
    12 -> 58 [ label = "(build)" ]
    19 -> 9 [ label = " 'cfg(unix)'" ]
    19 -> 59 [ label = " 'cfg(windows)'" ]
    19 -> 60 [ label = " 'cfg(windows)'" ]
    22 -> 43 [ label = "" ]
    22 -> 8 [ label = "" ]
    22 -> 61 [ label = "" ]
    22 -> 62 [ label = "" ]
    22 -> 63 [ label = "" ]
    23 -> 20 [ label = "" ]
    23 -> 64 [ label = "" ]
    23 -> 21 [ label = "" ]
    23 -> 65 [ label = "" ]
    26 -> 34 [ label = "" ]
    27 -> 26 [ label = "" ]
    27 -> 66 [ label = "" ]
    28 -> 29 [ label = "" ]
    28 -> 67 [ label = "" ]
    28 -> 68 [ label = "" ]
    28 -> 69 [ label = "" ]
    28 -> 70 [ label = "" ]
    28 -> 71 [ label = "" ]
    28 -> 72 [ label = "" ]
    28 -> 73 [ label = "" ]
    28 -> 74 [ label = "" ]
    30 -> 6 [ label = "(build)" ]
    30 -> 18 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    30 -> 75 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    30 -> 33 [ label = " 'cfg(any(target_arch = \"x86\", target_arch = \"x86_64\", all(any(target_arch = \"aarch64\", target_arch = \"arm\"), any(target_os = \"android\", target_os = \"fuchsia\", target_os = \"linux\"))))'" ]
    30 -> 35 [ label = "" ]
    30 -> 76 [ label = " 'cfg(target_os = \"windows\")'" ]
    30 -> 77 [ label = " 'cfg(target_os = \"windows\")'" ]
    38 -> 39 [ label = " 'i686-pc-windows-gnu'" ]
    38 -> 40 [ label = " 'x86_64-pc-windows-gnu'" ]
    58 -> 2 [ label = "" ]
    58 -> 45 [ label = "" ]
    43 -> 3 [ label = "" ]
    78 -> 4 [ label = "" ]
    78 -> 17 [ label = "" ]
    79 -> 4 [ label = "" ]
    79 -> 78 [ label = "" ]
    80 -> 4 [ label = "" ]
    80 -> 15 [ label = "" ]
    41 -> 4 [ label = "" ]
    41 -> 78 [ label = "" ]
    45 -> 10 [ label = "" ]
    45 -> 81 [ label = "" ]
    81 -> 10 [ label = "" ]
    81 -> 19 [ label = "" ]
    44 -> 10 [ label = "" ]
    44 -> 82 [ label = "" ]
    82 -> 10 [ label = "" ]
    82 -> 83 [ label = "" ]
    83 -> 10 [ label = "" ]
    83 -> 84 [ label = "" ]
    84 -> 10 [ label = "" ]
    84 -> 85 [ label = "" ]
    85 -> 10 [ label = "" ]
    85 -> 86 [ label = "" ]
    86 -> 10 [ label = "" ]
    86 -> 87 [ label = "" ]
    87 -> 10 [ label = "" ]
    87 -> 88 [ label = "" ]
    88 -> 10 [ label = "" ]
    50 -> 11 [ label = "" ]
    50 -> 89 [ label = "" ]
    50 -> 90 [ label = "" ]
    50 -> 91 [ label = "" ]
    50 -> 92 [ label = "" ]
    50 -> 93 [ label = "" ]
    93 -> 11 [ label = "" ]
    93 -> 56 [ label = "" ]
    92 -> 11 [ label = "" ]
    92 -> 57 [ label = "" ]
    91 -> 11 [ label = "" ]
    91 -> 55 [ label = "" ]
    90 -> 11 [ label = "" ]
    90 -> 54 [ label = "" ]
    89 -> 11 [ label = "" ]
    89 -> 53 [ label = "" ]
    57 -> 12 [ label = "" ]
    56 -> 12 [ label = "" ]
    55 -> 12 [ label = "" ]
    54 -> 12 [ label = "" ]
    53 -> 12 [ label = "" ]
    51 -> 13 [ label = "" ]
    62 -> 18 [ label = "" ]
    61 -> 18 [ label = "" ]
    61 -> 94 [ label = "" ]
    94 -> 18 [ label = "" ]
    64 -> 20 [ label = "" ]
    65 -> 21 [ label = "" ]
    52 -> 23 [ label = "" ]
    52 -> 95 [ label = "" ]
    52 -> 64 [ label = "" ]
    52 -> 65 [ label = "" ]
    95 -> 23 [ label = "" ]
    75 -> 24 [ label = "" ]
    75 -> 96 [ label = "" ]
    96 -> 24 [ label = "" ]
    96 -> 97 [ label = "" ]
    97 -> 24 [ label = "" ]
    66 -> 26 [ label = "" ]
    47 -> 28 [ label = "" ]
    47 -> 98 [ label = "" ]
    47 -> 99 [ label = "" ]
    47 -> 100 [ label = "" ]
    47 -> 101 [ label = "" ]
    47 -> 102 [ label = "" ]
    47 -> 103 [ label = "" ]
    47 -> 104 [ label = "" ]
    47 -> 67 [ label = "" ]
    104 -> 28 [ label = "" ]
    104 -> 74 [ label = "" ]
    103 -> 28 [ label = "" ]
    103 -> 73 [ label = "" ]
    102 -> 28 [ label = "" ]
    102 -> 72 [ label = "" ]
    101 -> 28 [ label = "" ]
    101 -> 71 [ label = "" ]
    100 -> 28 [ label = "" ]
    100 -> 70 [ label = "" ]
    99 -> 28 [ label = "" ]
    99 -> 69 [ label = "" ]
    98 -> 28 [ label = "" ]
    98 -> 68 [ label = "" ]
    46 -> 28 [ label = "" ]
    74 -> 29 [ label = "" ]
    73 -> 29 [ label = "" ]
    72 -> 29 [ label = "" ]
    71 -> 29 [ label = "" ]
    70 -> 29 [ label = "" ]
    69 -> 29 [ label = "" ]
    68 -> 29 [ label = "" ]
    67 -> 29 [ label = "" ]
    67 -> 68 [ label = "" ]
    67 -> 69 [ label = "" ]
    67 -> 70 [ label = "" ]
    67 -> 71 [ label = "" ]
    67 -> 72 [ label = "" ]
    67 -> 73 [ label = "" ]
    67 -> 74 [ label = "" ]
    42 -> 30 [ label = "" ]
    42 -> 105 [ label = "" ]
    42 -> 106 [ label = "" ]
    106 -> 30 [ label = "" ]
    106 -> 107 [ label = "" ]
    107 -> 30 [ label = "" ]
    107 -> 24 [ label = "" ]
    105 -> 30 [ label = "" ]
    48 -> 31 [ label = "" ]
    48 -> 108 [ label = "" ]
    108 -> 31 [ label = "" ]
    49 -> 32 [ label = "" ]
    49 -> 109 [ label = "" ]
    109 -> 32 [ label = "" ]
    63 -> 36 [ label = "" ]
    63 -> 110 [ label = "" ]
    110 -> 36 [ label = "" ]
    77 -> 38 [ label = "" ]
    76 -> 38 [ label = "" ]
    60 -> 38 [ label = "" ]
    59 -> 38 [ label = "" ]
}

